### DISCUSSION TITLE
How to use `TypeDirective`s?

### DISCUSSION THREAD

**Discussion by asimpleidea** [async-graphql/async-graphql#1444] [Q&A]
Hey everyone! 

Does anybody know how to use type directives? The documentation is not very helpful on this...
I currently have the following:

```rust
#[TypeDirective(
    location = "FieldDefinition",
    composable = "https://custom.spec.dev/extension/v1.0"
)]
pub fn testDirective(scope: String, input: u32, opt: Option<u64>) {
    println!("here")
}
```

And my object

```rust
#[derive(Debug, Clone, SimpleObject, Deserialize, Serialize)]
#[graphql(complex)]
pub struct MyTestObj {
    // Other objects...
    #[graphql(
        directive = testDirective::apply("field and param with \" symbol".to_string(), 2, Some(3))
    )]
    pub counter: u8,

}

#[ComplexObject]
impl MyTestObj {
    // Other things
}
```

The `testDirective` function is never executed, am I doing something wrong? Also, is there any way we can get the value of a field inside the function serving as a directive? Thank you!!

**Comment by Miaxos** [comment]
Type directives are used to add metadata on the schema itself, it can be used to communicate some metadata on a supergraph for instance[^1].

Here is the pull request where you'll be able to get more context on this: https://github.com/async-graphql/async-graphql/pull/1308

[^1]: https://www.apollographql.com/docs/federation/federated-types/federated-directives#managing-custom-directives

**Comment by asimpleidea** [comment]
Hi, yes I know how to use it theoretically, my question was more on how to use it async-graphql as I couldn't make it work, the function I was creating would never execute.

I will check the PR out and let you know if it works and mark your reply as answered in that case. Thanks!
